app/controllers/data_explorer/query_controller.rb

```diff
@@ -0,0 +1,236 @@
+    end
+
+    render_serialized queries, DataExplorer::QuerySerializer, root: 'queries'
+  end
+
+  skip_before_action :check_xhr, only: [:show]
+  def show
+    check_xhr unless params[:export]
+
+    query = DataExplorer::Query.find(params[:id])
+
+    if params[:export]
+      response.headers['Content-Disposition'] = "attachment; filename=#{query.slug}.dcquery.json"
+      response.sending_file = true
+    end
+
+    return raise Discourse::NotFound if !guardian.user_can_access_query?(query) || query.hidden
+    render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+  end
+
+  def groups
+    render_serialized(Group.all, BasicGroupSerializer)
+  end
+
+  def group_reports_index
+    return raise Discourse::NotFound unless guardian.user_is_a_member_of_group?(group)
+
+    respond_to do |format|
+      format.html { render 'groups/show' }
+      format.json do
+        queries = DataExplorer::Query
+          .where(hidden: false)
+          .joins("INNER JOIN data_explorer_query_groups data_explorer_query_groups 
+                    ON data_explorer_query_groups.query_id = data_explorer_queries.id 
+                    AND data_explorer_query_groups.group_id = #{group.id}")
+        render_serialized(queries, DataExplorer::QuerySerializer, root: 'queries')
+      end
+    end
+  end
+
+  def group_reports_show
+    return raise Discourse::NotFound if !guardian.group_and_user_can_access_query?(group, query) || query.hidden
+
+    respond_to do |format|
+      format.html { render 'groups/show' }
+      format.json do
+        render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+      end
+    end
+  end
+
+  skip_before_action :check_xhr, only: [:group_reports_run]
+  def group_reports_run
+    return raise Discourse::NotFound if !guardian.group_and_user_can_access_query?(group, query) || query.hidden
+
+    run
+  end
+
+  def create
+    query = DataExplorer::Query.create!(params.require(:query).permit(:name, :description, :sql).merge(user_id: current_user.id, last_run_at: Time.now).permit!)
+    group_ids = params.require(:query)[:group_ids]
+    group_ids&.each do |group_id|
+      query.query_groups.find_or_create_by!(group_id: group_id)
+    end
+    render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+  end
+
+  def update
+    query = DataExplorer::Query.find(params[:id])
+    query.update!(params.require(:query).permit(:name, :sql, :description).merge(hidden: false))
+
+    group_ids = params.require(:query)[:group_ids]
+    DataExplorer::QueryGroup.where.not(group_id: group_ids).delete_all
+    group_ids&.each do |group_id|
+      query.query_groups.find_or_create_by!(group_id: group_id)
+    end
+
+    render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+  rescue DataExplorer::ValidationError => e
+    render_json_error e.message
+  end
+
+  def destroy
+    query = DataExplorer::Query.where(id: params[:id]).first_or_initialize
```

Comment:
Hmm this doesn't make sense to me because this is a destroy action. If the `params[:id]` is not valid, we shouldn't have to create an instance of the object.

Reply:
yeah, I know it sounds weird but this is correct.

Besically what is happening here is that we have default queries in lib/queries.rb. However we need to give user an ability to remove them - hide them, similar to delete/hide custom queries saved in database.

So when user click "delete" button we create empty row in queries with hidden flag set to true and query id which is respected by index action